<html>
<head>
<title>How to Catch Multiple Exceptions in Python</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1 class="translated">如何在Python中捕捉多个异常</h1>
<blockquote>原文：<a href="https://towardsdatascience.com/multiple-exceptions-python-545a89c50d46#2022-04-27">https://towardsdatascience.com/multiple-exceptions-python-545a89c50d46#2022-04-27</a></blockquote><div><div class="fc ii ij ik il im"/><div class="in io ip iq ir"><div class=""/><div class=""><h2 id="1de9" class="pw-subtitle-paragraph jr it iu bd b js jt ju jv jw jx jy jz ka kb kc kd ke kf kg kh ki dk translated">在Python中处理多个异常</h2></div><figure class="kk kl km kn gu ko gi gj paragraph-image"><div role="button" tabindex="0" class="kp kq di kr bf ks"><div class="gi gj kj"><img src="../Images/5f57a2bd82648d6c278bc427bde45d44.png" data-original-src="https://miro.medium.com/v2/resize:fit:1400/format:webp/1*y5pWYXCh5rzBzG1TdUwW0g.jpeg"/></div></div><p class="kv kw gk gi gj kx ky bd b be z dk translated">照片由<a class="ae kz" href="https://unsplash.com/@chuttersnap?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText" rel="noopener ugc nofollow" target="_blank">丘特尔斯纳普</a>在<a class="ae kz" href="https://unsplash.com/s/photos/error?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText" rel="noopener ugc nofollow" target="_blank"> Unsplash </a>上拍摄</p></figure><h2 id="b572" class="la lb iu bd lc ld le dn lf lg lh dp li lj lk ll lm ln lo lp lq lr ls lt lu lv bi translated">介绍</h2><p id="e530" class="pw-post-body-paragraph lw lx iu ly b lz ma jv mb mc md jy me lj mf mg mh ln mi mj mk lr ml mm mn mo in bi translated">一个开发良好的应用程序必须总是能够以适当的方式处理意外事件——比如异常。这对于在开发期间调试源代码也很重要，对于在应用程序启动并运行(最终进入生产)时检查应用程序日志也很重要。</p><p id="c1fb" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">在今天的简短教程中，我们将展示如何在Python中处理多个异常。我们还将探索Python中的一些新特性，这些特性可以帮助您以一种更直观、更简洁的方式完成这项工作。我们开始吧！</p></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><h2 id="1904" class="la lb iu bd lc ld le dn lf lg lh dp li lj lk ll lm ln lo lp lq lr ls lt lu lv bi translated">对于Python &lt; 3.11</h2><p id="e40a" class="pw-post-body-paragraph lw lx iu ly b lz ma jv mb mc md jy me lj mf mg mh ln mi mj mk lr ml mm mn mo in bi translated">Now let’s suppose that we have the following (fairly dumb) code snippet, that raises  【T0】 ,  【T1】  and  【T2】 .</p><pre class="kk kl km kn gu nf ne ng nh aw ni bi"><span id="9bd8" class="la lb iu ne b gz nj nk l nl nm">def my_function(x):<br/>    if x == 1:<br/>        raise AttributeError('Example AttributeError')<br/>    elif x == 2:<br/>        raise ValueError('Example ValueError')<br/>    elif x == 3:<br/>        raise TypeError('Example TypeError')<br/>    else:<br/>        print('Hello World')</span></pre><p id="bdd6" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">And let’s also suppose that we want to to call the function  【T3】  but at the same time, we also need to ensure that we handle any unexpected errors appropriately. To do so, we can use the  【T4】  clauses.</p><p id="904c" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">But let’s assume that we want to perform a certain action if an  【T5】  is being thrown and a different action when either of  【T6】  or  【T7】  are raised by  【T8】 .</p><pre class="kk kl km kn gu nf ne ng nh aw ni bi"><span id="33d9" class="la lb iu ne b gz nj nk l nl nm">try:<br/>    my_function(x)<br/>except AttributeError:<br/>    # Do something<br/>    ...<br/>except (ValueError, TypeError):<br/>    # Do something else<br/>    ...</span></pre><blockquote class="nn no np"><p id="23d0" class="lw lx nq ly b lz mp jv mb mc mq jy me nr mr mg mh ns ms mj mk nt mt mm mn mo in bi translated">A  【T9】  statement may have more than one except clause, to specify handlers for different exceptions. At most one handler will be executed. Handlers only handle exceptions that occur in the corresponding try clause, not in other handlers of the same  【T10】  statement. An except clause may name multiple exceptions as a parenthesized tuple.</p><p id="752d" class="lw lx nq ly b lz mp jv mb mc mq jy me nr mr mg mh ns ms mj mk nt mt mm mn mo in bi translated">— <a class="ae kz" href="https://docs.python.org/2/tutorial/errors.html#handling-exceptions" rel="noopener ugc nofollow" target="_blank"> Python文档</a></p></blockquote><p id="d0b7" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">如果您不打算对引发的任何错误做任何特殊处理(例如，您只需<code class="fe nb nc nd ne b">pass</code>)，您甚至可以使用如下所示的<code class="fe nb nc nd ne b">suppress</code>上下文管理器:</p><pre class="kk kl km kn gu nf ne ng nh aw ni bi"><span id="f308" class="la lb iu ne b gz nj nk l nl nm">from contextlib import suppress<br/><br/>with suppress(<!-- -->AttributeError<!-- -->, <!-- -->ValueError, TypeError<!-- -->):<br/>     my_function(x)</span></pre><p id="cc62" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">注意<code class="fe nb nc nd ne b"><a class="ae kz" href="https://docs.python.org/3/library/contextlib.html#contextlib.suppress" rel="noopener ugc nofollow" target="_blank">suppress()</a></code>从Python 3.4开始可用。此外，只有当您希望程序的特定部分无声地失败并继续执行时，才必须使用这种方法。但是在大多数情况下，您可能希望对某些异常采取某些措施。</p></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><h2 id="a6d2" class="la lb iu bd lc ld le dn lf lg lh dp li lj lk ll lm ln lo lp lq lr ls lt lu lv bi translated">用Python 3.11处理多个异常</h2><p id="7ea8" class="pw-post-body-paragraph lw lx iu ly b lz ma jv mb mc md jy me lj mf mg mh ln mi mj mk lr ml mm mn mo in bi translated">从Python 3.11开始，引入了新的标准异常类型，即<code class="fe nb nc nd ne b">ExceptionGroup</code>。这个新异常用于<strong class="ly iv">一起传播一组不相关的异常</strong>。</p><p id="5f40" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">在下面的创建示例中，我们创建了一个包含四种不同类型错误的<code class="fe nb nc nd ne b">ExceptionGroup</code>实例，即<code class="fe nb nc nd ne b">TypeError</code>、<code class="fe nb nc nd ne b">ValueError</code>、<code class="fe nb nc nd ne b">KeyError</code>和<code class="fe nb nc nd ne b">AttributeError</code>。然后，我们使用多个<code class="fe nb nc nd ne b">except*</code>子句来处理<code class="fe nb nc nd ne b">ExceptionGroup</code>，无论是针对单个异常类型还是多个异常类型。</p><pre class="kk kl km kn gu nf ne ng nh aw ni bi"><span id="d0b1" class="la lb iu ne b gz nj nk l nl nm">try:<br/>    raise ExceptionGroup('Example ExceptionGroup', (<br/>        TypeError('Example TypeError'),<br/>        ValueError('Example ValueError'),<br/>        KeyError('Example KeyError'),<br/>        AttributeError('Example AttributeError')<br/>    ))<br/>except* TypeError:<br/>    ...<br/>except* ValueError as e:<br/>    ...<br/>except* (KeyError, AttributeError) as e:<br/>    ...</span></pre><p id="9a2f" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">但是请注意，一个<code class="fe nb nc nd ne b">except*</code>子句中提出的异常不适合匹配同一<code class="fe nb nc nd ne b">try</code>语句中的其他子句</p><p id="b40a" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">关于<code class="fe nb nc nd ne b">ExceptionGroup</code>和<code class="fe nb nc nd ne b">except*</code>条款背后的基本原理的更多细节，您可以参考<a class="ae kz" href="https://peps.python.org/pep-0654/" rel="noopener ugc nofollow" target="_blank"> PEP-654 </a>。</p><p id="a3de" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">此外，要更全面地阅读Python 3.11中的新增内容和更新，包括我们之前讨论的内容，您可以参考我最近在下面分享的一篇文章。</p><div class="nu nv gq gs nw nx"><a rel="noopener follow" target="_blank" href="/python-3-11-f62708eed569"><div class="ny ab fp"><div class="nz ab oa cl cj ob"><h2 class="bd iv gz z fq oc fs ft od fv fx it bi translated">Python 3.11有什么值得期待的</h2><div class="oe l"><h3 class="bd b gz z fq oc fs ft od fv fx dk translated">探索Python 3.11中的新增内容和更新，以及如何尽早获得3.11 Alpha版本</h3></div><div class="of l"><p class="bd b dl z fq oc fs ft od fv fx dk translated">towardsdatascience.com</p></div></div><div class="og l"><div class="oh l oi oj ok og ol kt nx"/></div></div></a></div></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><h2 id="47cd" class="la lb iu bd lc ld le dn lf lg lh dp li lj lk ll lm ln lo lp lq lr ls lt lu lv bi translated">最后的想法</h2><p id="f311" class="pw-post-body-paragraph lw lx iu ly b lz ma jv mb mc md jy me lj mf mg mh ln mi mj mk lr ml mm mn mo in bi translated">在今天的简短教程中，我们展示了在Python中处理多个异常的各种不同方法。我们已经看到了如何使用传统的<code class="fe nb nc nd ne b">except</code>子句来捕获多个异常，但是我们也展示了如何使用Python 3.11中将要引入的新的<code class="fe nb nc nd ne b">except*</code>子句来这样做。</p><p id="3ed9" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated">最后，您应该始终记住，在整个源代码中处理意外事件是一个重要的方面，如果执行得当，可以显著提高代码质量。</p></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><p id="1b77" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated"><a class="ae kz" href="https://gmyrianthous.medium.com/membership" rel="noopener"> <strong class="ly iv">成为会员</strong> </a> <strong class="ly iv">阅读介质上的每一个故事。你的会员费直接支持我和你看的其他作家。你也可以在媒体上看到所有的故事。</strong></p><div class="nu nv gq gs nw nx"><a href="https://gmyrianthous.medium.com/membership" rel="noopener follow" target="_blank"><div class="ny ab fp"><div class="nz ab oa cl cj ob"><h2 class="bd iv gz z fq oc fs ft od fv fx it bi translated">通过我的推荐链接加入Medium-Giorgos Myrianthous</h2><div class="oe l"><h3 class="bd b gz z fq oc fs ft od fv fx dk translated">作为一个媒体会员，你的会员费的一部分会给你阅读的作家，你可以完全接触到每一个故事…</h3></div><div class="of l"><p class="bd b dl z fq oc fs ft od fv fx dk translated">gmyrianthous.medium.com</p></div></div><div class="og l"><div class="om l oi oj ok og ol kt nx"/></div></div></a></div></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><p id="676e" class="pw-post-body-paragraph lw lx iu ly b lz mp jv mb mc mq jy me lj mr mg mh ln ms mj mk lr mt mm mn mo in bi translated"><strong class="ly iv">相关文章你可能也喜欢</strong></p><div class="nu nv gq gs nw nx"><a rel="noopener follow" target="_blank" href="/switch-statements-python-e99ea364fde5"><div class="ny ab fp"><div class="nz ab oa cl cj ob"><h2 class="bd iv gz z fq oc fs ft od fv fx it bi translated">如何用Python编写Switch语句</h2><div class="oe l"><h3 class="bd b gz z fq oc fs ft od fv fx dk translated">了解如何使用模式匹配或字典在Python中编写switch语句</h3></div><div class="of l"><p class="bd b dl z fq oc fs ft od fv fx dk translated">towardsdatascience.com</p></div></div><div class="og l"><div class="on l oi oj ok og ol kt nx"/></div></div></a></div></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><div class="kk kl km kn gu nx"><a rel="noopener follow" target="_blank" href="/how-to-merge-pandas-dataframes-221e49c41bec"><div class="ny ab fp"><div class="nz ab oa cl cj ob"><h2 class="bd iv gz z fq oc fs ft od fv fx it bi translated">如何合并熊猫数据帧</h2><div class="oe l"><h3 class="bd b gz z fq oc fs ft od fv fx dk translated">对熊猫数据帧执行左、右、内和反连接</h3></div><div class="of l"><p class="bd b dl z fq oc fs ft od fv fx dk translated">towardsdatascience.com</p></div></div><div class="og l"><div class="oo l oi oj ok og ol kt nx"/></div></div></a></div></div><div class="ab cl mu mv hy mw" role="separator"><span class="mx bw bk my mz na"/><span class="mx bw bk my mz na"/><span class="mx bw bk my mz"/></div><div class="in io ip iq ir"><div class="kk kl km kn gu nx"><a rel="noopener follow" target="_blank" href="/requirements-vs-setuptools-python-ae3ee66e28af"><div class="ny ab fp"><div class="nz ab oa cl cj ob"><h2 class="bd iv gz z fq oc fs ft od fv fx it bi translated">Python中的requirements.txt与setup.py</h2><div class="oe l"><h3 class="bd b gz z fq oc fs ft od fv fx dk translated">了解Python中requirements.txt、setup.py和setup.cfg在开发和分发时的用途…</h3></div><div class="of l"><p class="bd b dl z fq oc fs ft od fv fx dk translated">towardsdatascience.com</p></div></div><div class="og l"><div class="op l oi oj ok og ol kt nx"/></div></div></a></div></div></div>    
</body>
</html>